### PR TITLE
feat: support hook response body for ext-plugin

### PR DESCRIPTION
**Pull request by soulbird**
### Description

support hook response body for ext-plugin

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        host = ctx.picked_server.host,
+        port = ctx.picked_server.port,
+    })
+
+    if not ok then
+        return nil, err
+    end
+
+    local params = {
+        path = ctx.var.uri,
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:45]: Should use upstream_uri

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    return core.schema.check(_M.schema, conf)
+end
+
+
+function _M.response(conf, ctx)
+     -- TODO: request
+     local res, err = get_response(ctx)
+    if not res or err then
+        core.log.error("failed to request: ", err or "")
+        return core.response.exit(503)
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:81]: Use 502 would be better?

File: conf/config-default.yaml [hunk 3]
```diff
@@ -397,6 +397,7 @@ plugins:                          # plugin list (sorted by priority)
   - openwhisk                      # priority: -1901
   - serverless-post-function       # priority: -2000
   - ext-plugin-post-req            # priority: -3000
+  #- ext-plugin-post-resp          # priority: -4000
```
Comment (spacewander) [inline conf/config-default.yaml:400]: We can enable it by default?

File: apisix/plugins/ext-plugin/init.lua [hunk 4]
```diff
@@ -682,6 +683,96 @@ local rpc_handlers = {
 
         return true
     end,
+    nil,
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/init.lua:686]: Better to add a comment about the skipped key.

File: t/plugin/ext-plugin/response_body.t [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            ext.go({check_input = true})
+        }
+    }
+--- error_code: 200
+--- response_body
+hello world
+
+
+
+=== TEST 3: motify body
```
Comment (spacewander) [inline t/plugin/ext-plugin/response_body.t:125]: Need other test cases:
1. modify header
2. modify status
3. upstream failed to response
4. runner failed to response

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+     local headers = res.headers
+     local code, body = ext.communicate(conf, ctx, name, constants.RPC_HTTP_RESP_CALL)
+     if code or body then
+         -- TODO: chunk
+         return code, body
+     end
+     core.log.info("ext-plugin will send response")
+     -- send origin response
+     -- TODO: chunk
+     core.response.set_header(headers)
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:95]: We should hide some headers like Nginx's proxy_hide_header directive?

File: apisix/plugins/ext-plugin/init.lua [hunk 7]
```diff
@@ -682,6 +683,96 @@ local rpc_handlers = {
...
+        end
+
+        local buf = flatbuffers.binaryArray.New(resp)
+        local call_resp = http_resp_call_resp.GetRootAsResp(buf, 0)
+        local len = call_resp:HeadersLength()
+        if len > 0 then
+            local resp_headers = {}
+            for i = 1, len do
+                local entry = call_resp:Headers(i)
+                local name = str_lower(entry:Name())
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/init.lua:753]: We should exclude some headers like https://github.com/apache/apisix/blob/cc88a8db0ba64e1cb351e389ca5ec64c516073d6/apisix/plugins/ext-plugin/init.lua#L68?

File: conf/config-default.yaml (hunk 3 above)
Comment (soulbird) [reply conf/config-default.yaml:400]: All back-to-origin logic will be hooked by `resty.http`. So I think it is not appropriate to enable this plugin by default until the relevant logic is perfect.

Comment (spacewander) [reply conf/config-default.yaml:400]: Only route that contains this plugin should use `resty.http` instead.

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+
+    ngx.status = res.status
+
+    local reader = res.body_reader
+    repeat
+        local chunk, ok, read_err, print_err
+
+        chunk, read_err = reader()
+        if read_err then
+            core.error.log("read response failed: ", read_err)
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:161]: Ouch. We need tests to cover this branch.

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    -- Filter out exclude_resp_header headeres
+    for k, v in pairs(res.headers) do
+        if not exclude_resp_header[str_lower(k)] then
+            core.response.set_header(k, v)
+        end
+    end
+
+    local code, body = ext.communicate(conf, ctx, name, constants.RPC_HTTP_RESP_CALL)
+    if code or body then
+        -- TODO: chunk
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:147]: Why do we need to handle chunk in this branch?

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+                core.error.log("output response failed: ", print_err)
+            end
+        end
+
+        if read_err or print_err then
+            return 502
+        end
+    until not chunk
+
+    ngx_flush(true)
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:176]: The flush should be called after print, not here

File: t/plugin/ext-plugin/response_body.t [hunk 11]
```diff
@@ -0,0 +1,139 @@
+#
```
Comment (spacewander) [inline t/plugin/ext-plugin/response_body.t:1]: Should call it `response.t` as it is not only for the body

File: t/plugin/ext-plugin/response_body.t [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            ext.go({check_input = true})
+        }
+    }
+--- error_code: 200
+--- response_body
+hello world
+
+
+
+=== TEST 3: motify body
```
Comment (spacewander) [inline t/plugin/ext-plugin/response_body.t:123]:
```suggestion
=== TEST 3: modify body
```

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+    for k, v in pairs(res.headers) do
+        if not exclude_resp_header[str_lower(k)] then
+            core.response.set_header(k, v)
+        end
+    end
+
+    local code, body = ext.communicate(conf, ctx, name, constants.RPC_HTTP_RESP_CALL)
+    if code or body then
+        -- TODO: chunk
+        return code, body
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:148]: Should we close the reader here?

File: apisix/plugins/ext-plugin-post-resp.lua (hunk 10 above)
Comment (spacewander) [reply apisix/plugins/ext-plugin-post-resp.lua:176]: Also, need to log the err of ngx_flush if happened.

File: apisix/init.lua [hunk 14]
```diff
@@ -471,6 +481,9 @@ function _M.http_access_phase()
     if api_ctx.dubbo_proxy_enabled then
         return ngx.exec("@dubbo_pass")
     end
+
+    -- plugin runner response call
+    ext_response_phase("response")
```
[the remark this answers is not in the document]
Comment (spacewander) [reply apisix/init.lua:486]: We need to delay the `require` so the ext-plugin module won't be loaded if it is disabled. Or maybe there is a better way.

File: apisix/plugins/ext-plugin-post-resp.lua (hunk 13 above)
Comment (soulbird) [reply apisix/plugins/ext-plugin-post-resp.lua:131]: Now close first, keepalive will be processed later

File: t/plugin/ext-plugin/response.t [hunk 15]
```diff
@@ -0,0 +1,420 @@
...
+        }
+    }
+--- response_body
+passed
+
+
+
+=== TEST 2: check input
+--- request
+GET /hello
```
Comment (spacewander) [inline t/plugin/ext-plugin/response.t:107]: Is it possible to check if the args in the uri are sent to the upstream correctly?

File: apisix/plugins/ext-plugin/init.lua [hunk 16]
```diff
@@ -682,6 +682,107 @@ local rpc_handlers = {
...
+        if len > 0 then
+            -- TODO: support empty body
+            body = call_resp:BodyAsString()
+        end
+        local code = call_resp:Status()
+        core.log.info("recv resp, code: ", code, " body: ", body, " len: ", len)
+
+        if code == 0 then
+            -- runner change body only, we should set code.
+            code = body and res.status or nil
```
[the remark this answers is not in the document]
Comment (spacewander) [reply apisix/plugins/ext-plugin/init.lua:781]: Ping @soulbird

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+    end
+    core.log.info("ext-plugin will send response")
+
+    -- send origin response, status maybe changed.
+    err = send_response(res, code)
+    close(http_obj)
+
+    if err then
+        core.log.error(err)
+        return ngx.headers_sent and nil or 502
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:165]: 502 will always return here

File: apisix/plugins/ext-plugin-post-resp.lua [hunk 18]
```diff
@@ -162,7 +162,7 @@ function _M.before_proxy(conf, ctx)
 
     if err then
         core.log.error(err)
-        return ngx.headers_sent and nil or 502
+        return not ngx.headers_sent and 502
```
Comment (spacewander) [inline apisix/plugins/ext-plugin-post-resp.lua:165]: It would return `false` instead of `nil`?

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**